### ISSUE TITLE
Allow passing existing buffer to device.Read/ReadTimeout

### ISSUE THREAD

**Issue by TimonLukas** [badcel/HidApi.Net#101]
Hello! Thank you for building and publishing this library. I've been using it for a couple of days, and it's been a great experience :)

I'm using the library in an application which drives an externally connected USB device. In the hot path I'm calling `Device.ReadTimeout`. Since every call allocates a new array with `maxLength` bytes, this causes a lot of unnecessary GC pressure.

Would it be possible to add an overload/optional parameter for using an existing byte array instead of allocating a new one? This would allow for more predictable memory usage and improved performance, depending on the consuming application. It could for example be an overload replacing `maxLength`:

- `Device.Read(byte[] buffer)`
- `Device.ReadTimeout(byte[] buffer, int milliseconds)`

What do you think? If you like the idea, but are too busy, I'd be happy to open a PR.

Thank you!

**Comment by badcel** [comment]
Hi, thank you for your feedback. Nice that it works for you 👍 .

Yes I think it is a good idea to add such an overload. I would probably prefer to add either a `Span` or `ReadOnlySpan` instead of an `Array` as parameter. 

Those types are more flexible in regard of the memory location and are just a pointer and a length which is what I actually need to pass on. 

What would you prefer and why?

Do you use the `Write` method? I would assume having a `Span` instead of `ReadOnlySpan` would be better because it allows modification of the data which would make sense if the method is called in a loop for example to be able to update it.

If we agree on some way going forward you are welcome to create a PR.

**Comment by TimonLukas** [comment]
Thank you for the very quick response!

`Span` should work perfectly. I tried to look into whether `Memory` might also make sense, but if I understand [rule 9 on this page](https://learn.microsoft.com/en-us/dotnet/standard/memory-and-spans/memory-t-usage-guidelines) correctly, `Span`s would be preferred in this case.

I do use `Write`, but with separate buffers. The application I'm writing is meant to control Stream Deck devices (with the HID logic ported from [`Julusian/node-elgato-stream-deck`](https://github.com/Julusian/node-elgato-stream-deck/tree/main)). To keep things simple, the main loop consists of two simultaneous `PeriodTimer` tasks:
- The reading one polls the device (using blocking mode with a small timeout), maps the read data to input events (e.g. button down/up) and emits them
    - This is probably not the best approach, it would be better to poll on a separate thread, but I wanted to keep things simple to start with, and this provides great latency :)
- The updating/writing one both handles device commands (set brightness etc.), as well as rendering and writing new images

So `Read`/`Write` calls are pretty far apart in my application.

I still have a lot of cleaning up to do before I'll make the repo public (as I'm still pretty inexperienced with C#), but in case you're interested I added you as a collaborator. Here are examples for the calls:
- [`ReadTimeout`](https://github.com/TimonLukas/dotnet-stream-deck-controller/blob/cf5d62dae9e7bae95d62ac2b600ffd25f27fb5cf/StreamDeckController/StreamDeckApi/Models/BaseInput.cs#L110)
- [`Write`](https://github.com/TimonLukas/dotnet-stream-deck-controller/blob/cf5d62dae9e7bae95d62ac2b600ffd25f27fb5cf/StreamDeckController/StreamDeckApi/Models/Base.cs#L164)
- [`SendFeatureReport`](https://github.com/TimonLukas/dotnet-stream-deck-controller/blob/cf5d62dae9e7bae95d62ac2b600ffd25f27fb5cf/StreamDeckController/StreamDeckApi/Models/Base.cs#L46)
- [`GetFeatureReport`](https://github.com/TimonLukas/dotnet-stream-deck-controller/blob/cf5d62dae9e7bae95d62ac2b600ffd25f27fb5cf/StreamDeckController/StreamDeckApi/Models/BaseGen2.cs#L47)

By the way, thank you especially for the great AoT support! I had expected at least some setup difficulties, but it worked perfectly under Linux without any extra setup.

**Comment by badcel** [comment]
Thanks I will take a look 👍 

I think my suggestion would be:
- Add new methods to read data (regular + with timeout) which accept a `Span` as parameter.
- The original "read" methods should probably call into the newly created ones as they are just providing automatic memory allocation on top of the new ones.
- The internal methods probably have to switch from `ReadOnlySpan` to `Span`, too.

As the`Write` method is already using `ReadOnlySpan` in the public API I would probably don't add an additional overload until there is demand for it as it would require creating a new call hierarchy which is identical to the existing one. If needed it could perhaps replace the existing one, forcing a major version bump. I just don't have enough information what is needed but as it works for you it can't be unusable.

If you like you can create the PR for the "read" methods. I would then go ahead and create a release in the coming days.

Sidenote: While I read through some of the `Span` documentation yesterday I came across the point that it can't be used in `async` code. If you require such a feature for your application we can take a look into it if the need arises.

**Comment by TimonLukas** [comment]
Good point about `Span`s and `async` - I ran into the mentioned issue when I started the project, which made me use my custom byte-array wrapper class everywhere. This works fine for simple scenarios, but since I'm basically circumventing the C# approach to safe memory management, I'll refactor the application to follow the [`Memory<T>` and `Span<T>` usage guidelines](https://learn.microsoft.com/en-us/dotnet/standard/memory-and-spans/memory-t-usage-guidelines) (essentially switching to `Memory<T>`, but with proper ownership).

Based on these guidelines I'd propose a couple of extensions to what you suggested:

> I think my suggestion would be:
> * Add new methods to read data (regular + with timeout) which accept a `Span` as parameter.

This will be useful, but as you mentioned, this wouldn't work well in `async` contexts (or it would at least require boilerplate workarounds every time). Maybe the best option is just to offer overloads for all scenarios:

```csharp
public interface IDevice {
    ReadOnlySpan<byte> Read(int maxLength);
    Span<byte> Read(Span<byte> buffer);
    Memory<byte> Read(Memory<byte> buffer);
    // If you want to offer max. flexibility
    ReadOnlySpan<byte> Read(byte[] buffer);

    ReadOnlySpan<byte> ReadTimeout(int maxLength, int milliseconds);
    Span<byte> ReadTimeout(Span<byte> buffer, int milliseconds);
    Memory<byte> ReadTimeout(Memory<byte> buffer, int milliseconds);
    // If you want to offer max. flexibility
    ReadOnlySpan<byte> ReadTimeout(byte[] buffer, int milliseconds);
}
```

This should cover every case that will realistically happen.

> * The original "read" methods should probably call into the newly created ones as they are just providing automatic memory allocation on top of the new ones.

Sounds like a good plan! I'm not sure if this approach could be done for all overloads, `Memory<byte>` might have to be implemented separately, but I'll experiment a bit to see what's possible.

> * The internal methods probably have to switch from `ReadOnlySpan` to `Span`, too.

Yep!

> As the `Write` method is already using `ReadOnlySpan` in the public API I would probably don't add an additional overload until there is demand for it as it would require creating a new call hierarchy which is identical to the existing one. [...]

`Write` is totally fine as it is, as it's only a consumer, and most "kinds of memory" have implicit conversions to `ReadOnlySpan`. The only possible addition I can think of would be a method that accepts a `ReadOnlyMemory` and passes the `ReadOnlySpan` along:

```csharp
public interface IDevice {
    void Write(ReadOnlySpan<byte> data);
    
    void Write(ReadOnlyMemory<byte> memory) {
        Write(memory.Span);
    }
```

> If you like you can create the PR for the "read" methods. I would then go ahead and create a release in the coming days.

I'll gladly give it a try! Before submitting, I'd definitely like to experiment with `Memory` a bit. Hopefully I should be finished with experiments & PR by end of next week :) (depending on your further feedback of course!)

**Comment by badcel** [comment]
Let's keep `Span` and `Memory` in separate PRs. I think `Span` is quite uncontroversial.

I would like to take a deeper look into the changes related to `Memory`. Perhaps you propose a new API before implementing it like you did here for `Span` after collecting some experience with it.

Regarding your proposed API I would probably see it a little different:

- Read without a buffer parameter returns a `Span` so there is no conversion needed between `Span` and `ReadOnlySpan` if it is based on the method with a buffer parameter.
- Read with buffer returns void, as the buffer is filled.
- No overload with `byte[]` needed as there is an implicit conversion available from an array to `Span`:

```csharp
public interface IDevice {
    Span<byte> Read(int maxLength);
    void Read(Span<byte> buffer);
    Span<byte> ReadTimeout(int maxLength, int milliseconds);
    void ReadTimeout(Span<byte> buffer, int milliseconds);
}
```

(Please don't add an `IDevice` interface, this would be another discussion.)

What do you think of my api suggestion?

**Comment by badcel** [comment]
We could even consider removing the API which returns `Span` as it is only a helper method and only keep the ones which return `void` and have a `Span` as parameter.

I would like to keep the API as simple as possible. This change would move into this direction and would justify a major version bump. The bump indicates the majority of the API and it would be a lot more flexible than before:

```csharp
public interface IDevice {
    void Read(Span<byte> buffer);
    void ReadTimeout(Span<byte> buffer, int milliseconds);
}
```

**Comment by TimonLukas** [comment]
> Let's keep `Span` and `Memory` in separate PRs. I think `Span` is quite uncontroversial.
> I would like to take a deeper look into the changes related to Memory. Perhaps you propose a new API before implementing it like you did here for Span after collecting some experience with it.

Sounds good!

> We could even consider removing the API which returns Span as it is only a helper method and only keep the ones which return void and have a Span as parameter.

This is an option, but there is also value in the simplicity of the current methods. For the average case, the allocations aren't an issue, and being sure the `ReadOnlySpan` value won't change under your nose makes it very easy to reason about. The `Span` approach is easier to misuse (e.g. inconsistent writes).

I'm generally a fan of both options being offered - one "a little wasteful, but impossible to misuse", and one "higher performance, at your own risk". But I'll stay a happy user either way, you have to decide :)

> * Read without a buffer parameter returns a Span so there is no conversion needed between Span and ReadOnlySpan if it is based on the method with a buffer parameter.

There is an advantage to no-buffer-`Read` returning a `ReadOnlySpan`, as it communicates that the result will not change (and it's usable in places where `Span` isn't). Since `Span<T>` has an implicit conversion to `ReadOnlySpan<T>`, there shouldn't be any extra code required. I'd suggest that I'll implement both options, and you can decide whether the additional complexity is worth it based on how common scenarios could be implemented?

> * Read with buffer returns void, as the buffer is filled.

This makes sense, as it communicates more clearly what is going on.

> * No overload with `byte[]` needed as there is an implicit conversion available from an array to `Span`:

You're absolutely right!

> (Please don't add an `IDevice` interface, this would be another discussion.)

Just a placeholder for discussing the design, no worries :)

> What do you think of my api suggestion?

I like it a lot! Based on the points above I'd make the slight change of still returning a `ReadOnlySpan` for the non-buffer-versions:

```csharp
public interface IDevice {
    ReadOnlySpan<byte> Read(int maxLength);
    void Read(Span<byte> buffer);
    ReadOnlySpan<byte> ReadTimeout(int maxLength, int milliseconds);
    void ReadTimeout(Span<byte> buffer, int milliseconds);
}
```

This would mean that no major version bump is necessary, and it would clearly communicate which version of the method should be used in which case. But it's your call, removing the non-buffer-versions would work just as well for me :)

**Comment by badcel** [comment]
I thought a little bit about your arguments and I prefer to remove the current method signature in favor of the new with a buffer parameter.

If a user of HidApi.Net wants to have this method for its simplicity it is a easy to implement as an extension method.

At the same time I don't want to have it as part of HidApi.Net because it might not be the optimal implementation. You came here to reduce gc pressure. This is probably not a problem for spans as they don't live on the heap. But a user might already have some native / managed memory available or even prefer `stackalloc`.

Adding one helper method would bring up the question why there is not another one for `stackalloc` or why `stackalloc` is not automatically used if the buffer is small enough. Avoiding this kind of intricate implementation is kind of easy if there is only one api providing maximum flexibility and requires some thought of the user. But I consider this a good thing actually.

**Comment by TimonLukas** [comment]
Makes sense! I've updated the PR accordingly (and fixed the doc comments). It should be ready for review!

I only updated the signatures of the native methods that are used by `Read` / `ReadTimeout` to use `Span`s instead of `ReadOnlySpan`s, though this should probably also be done for the other `Get` methods which take buffers, e.g. `GetInputReport`. Would you like me to open a separate issue/PR for that? :)

**Comment by badcel** [comment]
What you did is exactly right as it is what we discussed 👍.

The other "Get" functions are a different problem(s). If you have some suggestion please open a new issue and let's discuss it first.

**Comment by badcel** [comment]
Implemented in #102.
